File: mcp-code-intelligence-python/src/mcp_code_intel/memory/quality_gate.py
"""QualityGate — validates content before ingest to prevent KB pollution.

Checks: minimum length, duplicate detection (trigram Jaccard), quality scoring.
Rejects entries with score < 30, warns for score 30-50.
"""

import re
import sqlite3
from dataclasses import dataclass


@dataclass
class QualityResult:
    """Result of quality validation."""

    score: int
    decision: str  # "accept", "warn", "reject"
    message: str | None
    duplicate_detected: bool
    duplicate_entry_id: int | None


@dataclass
class IngestMeta:
    """Metadata provided during ingest for quality scoring."""

    tags: str | None = None
    type: str | None = None
    source: str | None = None


class QualityGate:
    """Validates content quality before KB ingest."""

    def __init__(
        self,
        conn: sqlite3.Connection,
        min_length: int = 50,
        reject_threshold: int = 30,
        warn_threshold: int = 50,
        duplicate_threshold: float = 0.95,
    ) -> None:
        self._conn = conn
        self._min_length = min_length
        self._reject_threshold = reject_threshold
        self._warn_threshold = warn_threshold
        self._duplicate_threshold = duplicate_threshold

    def validate(self, content: str, meta: IngestMeta) -> QualityResult:
        """Validate content before ingest. Returns quality decision."""
        trimmed = content.strip()
        if len(trimmed) < self._min_length:
            return QualityResult(
                score=0, decision="reject",
                message=f"Content too short (min {self._min_length} chars). Got {len(trimmed)}.",
                duplicate_detected=False, duplicate_entry_id=None,
            )

        sim, dup_id = self._check_duplicate(trimmed)
        if sim >= self._duplicate_threshold:
            return QualityResult(
                score=10, decision="reject",
                message=f"Duplicate detected (similarity: {sim * 100:.1f}%). Existing entry ID: {dup_id}",
                duplicate_detected=True, duplicate_entry_id=dup_id,
            )

        score = self._calculate_score(trimmed, meta)
        decision = self._decide_from_score(score)
        message = (
            f"Low quality score ({score}/100). Consider adding tags, source, or more detail."
            if decision == "warn" else None
        )
        return QualityResult(score, decision, message, False, None)

    def _calculate_score(self, content: str, meta: IngestMeta) -> int:
        score = 0
        score += min(40, (len(content) * 40) // 500)
        if meta.tags and meta.tags.strip():
            score += 20
        if meta.type and meta.type.strip():
            score += 10
        if meta.source and meta.source.strip():
            score += 10
        if _has_structure(content):
            score += 10
        if _has_actionable_content(content):
            score += 10
        return min(100, score)

    def _check_duplicate(self, content: str) -> tuple[float, int | None]:
        trigrams = _build_trigrams(content.lower()[:200])
        if not trigrams:
            return 0.0, None

        cursor = self._conn.cursor()
        cursor.execute(
            "SELECT id, content FROM knowledge_entries "
            "WHERE archived = 0 ORDER BY created_at DESC LIMIT 50"
        )
        max_sim = 0.0
        match_id: int | None = None

        for row in cursor.fetchall():
            cand_trigrams = _build_trigrams(row[1].lower()[:200])
            sim = _jaccard_similarity(trigrams, cand_trigrams)
            if sim > max_sim:
                max_sim = sim
                match_id = row[0]
        return max_sim, match_id

    def _decide_from_score(self, score: int) -> str:
        if score < self._reject_threshold:
            return "reject"
        if score < self._warn_threshold:
            return "warn"
        return "accept"


def _has_structure(text: str) -> bool:
    return bool(
        re.search(r"^#{1,6}\s", text, re.MULTILINE)
        or re.search(r"^[-*]\s", text, re.MULTILINE)
        or "```" in text
    )


def _has_actionable_content(text: str) -> bool:
    return bool(
        re.search(r"\b(TODO|Action:|Decision:|Next step:|Decided:)\b", text, re.IGNORECASE)
        or re.search(r"\[[ x]\]", text, re.IGNORECASE)
    )


def _build_trigrams(text: str) -> set[str]:
    return {text[i:i + 3] for i in range(len(text) - 2)}


def _jaccard_similarity(a: set[str], b: set[str]) -> float:
    if not a and not b:
        return 1.0
    intersection = len(a & b)
    union = len(a) + len(b) - intersection
    return intersection / union if union else 0.0

File: mcp-code-intelligence-kotlin/src/main/kotlin/com/codeintel/memory/search/QualityGate.kt
/**
 * QualityGate — validates content before ingest to prevent KB pollution.
 * Checks: minimum length, duplicate detection (trigram Jaccard), quality scoring.
 * Rejects entries with score < 30, warns for score 30-50.
 */
package com.codeintel.memory.search

import java.sql.Connection

/** Result of quality validation. */
data class QualityResult(
    val score: Int,
    val decision: String, // "accept", "warn", "reject"
    val message: String?,
    val duplicateDetected: Boolean,
    val duplicateEntryId: Long?
)

/** Metadata provided during ingest for quality scoring. */
data class IngestMeta(
    val tags: String? = null,
    val type: String? = null,
    val source: String? = null
)

class QualityGate(
    private val conn: Connection,
    private val minLength: Int = 50,
    private val rejectThreshold: Int = 30,
    private val warnThreshold: Int = 50,
    private val duplicateThreshold: Double = 0.95
) {

    /** Validate content before ingest. Returns quality decision. */
    fun validate(content: String, meta: IngestMeta): QualityResult {
        val trimmed = content.trim()
        if (trimmed.length < minLength) {
            return QualityResult(
                score = 0, decision = "reject",
                message = "Content too short (min $minLength chars). Got ${trimmed.length}.",
                duplicateDetected = false, duplicateEntryId = null
            )
        }

        val dup = checkDuplicate(trimmed)
        if (dup.first >= duplicateThreshold) {
            return QualityResult(
                score = 10, decision = "reject",
                message = "Duplicate detected (similarity: ${"%.1f".format(dup.first * 100)}%). Existing entry ID: ${dup.second}",
                duplicateDetected = true, duplicateEntryId = dup.second
            )
        }

        val score = calculateScore(trimmed, meta)
        val decision = decideFromScore(score)
        val message = if (decision == "warn")
            "Low quality score ($score/100). Consider adding tags, source, or more detail."
        else null

        return QualityResult(score, decision, message, false, null)
    }

    private fun calculateScore(content: String, meta: IngestMeta): Int {
        var score = 0
        score += minOf(40, (content.length * 40) / 500)
        if (!meta.tags.isNullOrBlank()) score += 20
        if (!meta.type.isNullOrBlank()) score += 10
        if (!meta.source.isNullOrBlank()) score += 10
        if (hasStructure(content)) score += 10
        if (hasActionableContent(content)) score += 10
        return minOf(100, score)
    }

    private fun checkDuplicate(content: String): Pair<Double, Long?> {
        val trigrams = buildTrigrams(content.lowercase().take(200))
        if (trigrams.isEmpty()) return 0.0 to null

        val candidates = loadRecentEntries()
        var maxSim = 0.0
        var matchId: Long? = null

        for ((id, candidateContent) in candidates) {
            val candidateTrigrams = buildTrigrams(candidateContent.lowercase().take(200))
            val sim = jaccardSimilarity(trigrams, candidateTrigrams)
            if (sim > maxSim) { maxSim = sim; matchId = id }
        }
        return maxSim to matchId
    }

    private fun loadRecentEntries(): List<Pair<Long, String>> {
        val results = mutableListOf<Pair<Long, String>>()
        conn.prepareStatement(
            "SELECT id, content FROM knowledge_entries WHERE archived = 0 ORDER BY created_at DESC LIMIT 50"
        ).use { stmt ->
            stmt.executeQuery().use { rs ->
                while (rs.next()) results.add(rs.getLong("id") to rs.getString("content"))
            }
        }
        return results
    }

    private fun decideFromScore(score: Int): String = when {
        score < rejectThreshold -> "reject"
        score < warnThreshold -> "warn"
        else -> "accept"
    }
}

private fun hasStructure(text: String): Boolean =
    Regex("^#{1,6}\\s", RegexOption.MULTILINE).containsMatchIn(text) ||
        Regex("^[-*]\\s", RegexOption.MULTILINE).containsMatchIn(text) ||
        "```" in text

private fun hasActionableContent(text: String): Boolean =
    Regex("\\b(TODO|Action:|Decision:|Next step:|Decided:)", RegexOption.IGNORE_CASE)
        .containsMatchIn(text) || Regex("\\[[ x]]", RegexOption.IGNORE_CASE).containsMatchIn(text)

private fun buildTrigrams(text: String): Set<String> {
    val set = mutableSetOf<String>()
    for (i in 0..text.length - 3) set.add(text.substring(i, i + 3))
    return set
}

private fun jaccardSimilarity(a: Set<String>, b: Set<String>): Double {
    if (a.isEmpty() && b.isEmpty()) return 1.0
    val intersection = a.count { it in b }
    val union = a.size + b.size - intersection
    return if (union == 0) 0.0 else intersection.toDouble() / union
}
